core-primitives/attestation-handler/src/attestation_handler.rs

```diff
@@ -0,0 +1,661 @@
+//    the documentation and/or other materials provided with the
+//    distribution.
+//  * Neither the name of Baidu, Inc., nor the names of its
+//    contributors may be used to endorse or promote products derived
+//    from this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+// "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+// LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+// A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+// OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+// LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+// DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+// THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+#[cfg(all(not(feature = "std"), feature = "sgx"))]
+use crate::sgx_reexport_prelude::*;
+
+use crate::{cert, Error as EnclaveError, Error, Result as EnclaveResult};
+use codec::Encode;
+use core::default::Default;
+use itertools::Itertools;
+use itp_hashing::hash_from_slice;
+use itp_node_api::{
+	api_client::{ParentchainExtrinsicParams, ParentchainExtrinsicParamsBuilder},
+	metadata::{pallet_teerex::TeerexCallIndexes, provider::AccessNodeMetadata, NodeMetadata},
+};
+use itp_nonce_cache::{MutateNonce, Nonce, GLOBAL_NONCE_CACHE};
+use itp_ocall_api::EnclaveAttestationOCallApi;
+use itp_settings::{
+	files::{RA_API_KEY_FILE, RA_DUMP_CERT_DER_FILE, RA_SPID_FILE},
+	worker::MR_ENCLAVE_SIZE,
+};
+use itp_sgx_crypto::Ed25519Seal;
+use itp_sgx_io as io;
+use itp_sgx_io::StaticSealedIO;
+use log::*;
+use sgx_rand::{os, Rng};
+use sgx_tcrypto::{rsgx_sha256_slice, SgxEccHandle};
+use sgx_tse::{rsgx_create_report, rsgx_verify_report};
+use sgx_types::{
+	c_int, sgx_epid_group_id_t, sgx_quote_nonce_t, sgx_quote_sign_type_t, sgx_report_data_t,
+	sgx_spid_t, sgx_status_t, sgx_target_info_t, SgxResult,
+};
+use sp_core::{blake2_256, Pair};
+use std::{
+	borrow::ToOwned,
+	format,
+	io::{Read, Write},
+	net::TcpStream,
+	println, str,
+	string::{String, ToString},
+	sync::Arc,
+	vec::Vec,
+};
+use substrate_api_client::{compose_extrinsic_offline, ExtrinsicParams};
+
+pub const DEV_HOSTNAME: &str = "api.trustedservices.intel.com";
+
+#[cfg(feature = "production")]
+pub const SIGRL_SUFFIX: &str = "/sgx/attestation/v4/sigrl/";
+#[cfg(feature = "production")]
+pub const REPORT_SUFFIX: &str = "/sgx/attestation/v4/report";
+
+#[cfg(not(feature = "production"))]
+pub const SIGRL_SUFFIX: &str = "/sgx/dev/attestation/v4/sigrl/";
+#[cfg(not(feature = "production"))]
+pub const REPORT_SUFFIX: &str = "/sgx/dev/attestation/v4/report";
+
+pub trait AttestationHandler {
```

Comment:
Could you add some documentation to this trait? No need at this point to do it for the implementing structs, but the public facing traits should have some minimal documentation / comment.

Reply:
You are right the attestation code could benefit from some better documentation and comments. I added some minimal docs to the trait. 
I will try to improve on it in the future.